scipy/sparse/tests/test_base.py

```diff
@@ -1146,9 +1146,16 @@ def test_astype(self):
         S = self.spmatrix(D)
 
         for x in supported_dtypes:
-            assert_equal(S.astype(x).dtype, D.astype(x).dtype)  # correct type
-            assert_equal(S.astype(x).toarray(), D.astype(x))        # correct values
-            assert_equal(S.astype(x).format, S.format)           # format preserved
+            # Check correct casted
+            D_casted = D.astype(x)
+            for copy in (True, False):
+                S_casted = S.astype(x, copy=copy)
+                assert_equal(S_casted.dtype, D_casted.dtype)  # correct type
+                assert_equal(S_casted.toarray(), D_casted)    # correct values
+                assert_equal(S_casted.format, S.format)       # format preserved
+            # Check correct copied
+            assert S_casted.astype(x, copy=False) is S_casted
+            assert S_casted.astype(x, copy=True) is not S_casted
```

Comment:
This doesn't necessarily catch the indices-sharing bug: the sparse matrix objects may be distinct but still share memory for some or all member arrays. To be absolutely sure, we could do something like:
```python
A = self.spmatrix(D)
B = A.astype(float, copy=True)  #  A.dtype is complex, so this must cast
assert_array_equal(A.indices, B.indices)
A.indices[0] += 1
assert_(A.indices[0] != B.indices[0])
```